aws/image/misc-settings.sh

```diff
@@ -20,24 +20,95 @@ if [ "$CLOUD_PROVIDER" == "vsphere" ]
-		(! dnf -y install open-vm-tools) && \
-		     echo "$PODVM_DISTRO: Error installing package required for cloud provider: $CLOUD_PROVIDER" 1>&2 && exit 1
-	    ;;
-	ubuntu)
-	    (! dpkg -l | grep open-vm-tools > /dev/null 2>&1) && apt-get update && \
-		(! apt-get -y install open-vm-tools > /dev/null 2>&1) && \
-		     echo "$PODVM_DISTRO: Error installing package required for cloud provider: $CLOUD_PROVIDER" 1>&2  && exit 1
-	    ;;
-	*)
-	    ;;
+    centos)
+        #fallthrough
+        ;&
+    rhel)
+        (! dnf list --installed | grep open-vm-tools > /dev/null 2>&1) && \
+        (! dnf -y install open-vm-tools) && \
+             echo "$PODVM_DISTRO: Error installing package required for cloud provider: $CLOUD_PROVIDER" 1>&2 && exit 1
+        ;;
+    ubuntu)
+        (! dpkg -l | grep open-vm-tools > /dev/null 2>&1) && apt-get update && \
+        (! apt-get -y install open-vm-tools > /dev/null 2>&1) && \
+             echo "$PODVM_DISTRO: Error installing package required for cloud provider: $CLOUD_PROVIDER" 1>&2  && exit 1
+        ;;
+    *)
+        ;;
     esac
-# else if...
-#
+fi
+
+# Setup oneshot systemd service for AWS and Azure to enable NAT rules
+if [ "$CLOUD_PROVIDER" == "azure" ] || [ "$CLOUD_PROVIDER" == "aws" ]
+then
+# Create a oneshot systemd service to execute setup-nat-for-imds.sh
+# during first boot
+cat <<END > /etc/systemd/system/setup-nat-for-imds.service
+[Unit]
+Description=Setup NAT for IMDS
+After=kata-agent.service
+Wants=kata-agent.service
+
+[Service]
+Type=oneshot
+ExecStart=/usr/local/bin/setup-nat-for-imds.sh
+
+[Install]
+WantedBy=multi-user.target
+END
+
+# Add a link to the service in multi-user.target.wants
+ln -s /etc/systemd/system/setup-nat-for-imds.service /etc/systemd/system/multi-user.target.wants/setup-nat-for-imds.service
+
+# Create a script to setup NAT for IMDS
+cat <<END > /usr/local/bin/setup-nat-for-imds.sh
+#!/bin/bash
+# This script sets up NAT for IMDS on Azure/AWS
+# This is required for IMDS to work on Azure/AWS
+# This script is executed as a oneshot systemd service
+# during first boot
+
+set -euo pipefail
+
+IMDS_IP="169.254.169.254"
+DUMMY_IP="169.254.99.99"
+
+# trap errors
+trap 'echo "Error: \$0:\$LINENO stopped"; exit 1' ERR INT
+
+# Function to setup veth pair
+function setup_proxy_arp() {
+  local pod_ip=${pod_ip:-}
+  pod_ip=$(ip netns exec podns ip route get "$IMDS_IP" | awk 'NR == 1 { print $7 }')
+
+  ip link add veth2 type veth peer name veth1
+  ip address add "$DUMMY_IP/32" dev veth1 # Proxy arp does not get enabled when no IP address is assigned
+  ip link set up dev veth1
+
+  sysctl -w net.ipv4.ip_forward=1
+  sysctl -w net.ipv4.conf.veth1.proxy_arp=1
+  sysctl -w net.ipv4.neigh.veth1.proxy_delay=0
+
+  ip link set veth2 netns podns
+  ip netns exec podns ip link set up dev veth2
+  ip netns exec podns ip route add "$IMDS_IP/32" dev veth2
+
+  ip route add "$pod_ip/32" dev veth1
+
+  local hwaddr=${hwaddr:-}
```

Comment:
Same as above.

Reply:
This was to make shellcheck happy. Otherwise it threw error.

Reply:
Oh, shellcheck examines the shell script contained in the heredoc (`cat <<END`)?

I also noticed that `$IMDS_IP` and `$DUMMY_IP` will be expanded to empty values when the `setup-nat-for-imds.sh` script file is generated.

I think we need to surround `END` with a pair of single quote `'` like this to suppress such variable substation in the heredoc.
```
cat <<'END' > /usr/local/bin/setup-nat-for-imds.sh
```

Reply:
The `$` needs to be escaped so that it's not expanded.

Reply:
Here is an explanation in shellcheck. 
https://github.com/koalaman/shellcheck/issues/2409#issuecomment-990441548

I am not sure why `local pod_ip=${pod_ip:-}` suppresses the shell error, but I think it does not fix the issue.

Use of `cat <<'END'` will suppress expansion of any variables in the here doc.

Reply:
> I also noticed that $IMDS_IP and $DUMMY_IP will be expanded to empty values when the setup-nat-for-imds.sh script file is generated.

Shellcheck assumes that shell vars in caps are external to the script, while lowercase variables are internal to the script (some sort of convention I guess). That's the reason that `hwaddr` is linted when not defines in the script, but shellcheck doesn't complain about `IMDS_IP` when it is not defined.

Reply:
> I am not sure why local pod_ip=${pod_ip:-} suppresses the shell error

It sets a default value (behind the `:-` , so the default is explicitly empty), so `pod_ip` is initialized.

Reply:
Updated the PR with the suggested changes